File: test/extensions/language/offsets/java/offsets/OptionalMemberOffsetTest.java
package offsets;

import static org.junit.jupiter.api.Assertions.*;

import java.io.IOException;

import org.junit.jupiter.api.Test;

import zserio.runtime.ZserioError;
import zserio.runtime.io.BitBuffer;
import zserio.runtime.io.BitStreamReader;
import zserio.runtime.io.ByteArrayBitStreamReader;
import zserio.runtime.io.ByteArrayBitStreamWriter;
import zserio.runtime.io.SerializeUtil;

import offsets.optional_member_offset.OptionalMemberOffset;

public class OptionalMemberOffsetTest
{
    @Test
    public void readWithOptional() throws IOException, ZserioError
    {
        final boolean hasOptional = true;
        final int optionalField = 0x1212;
        final int field = 0x2121;
        final BitBuffer bitBuffer =
                writeOptionalMemberOffsetToBitBuffer(hasOptional, OPTIONAL_FIELD_OFFSET, optionalField, field);
        final BitStreamReader reader = new ByteArrayBitStreamReader(bitBuffer);
        final OptionalMemberOffset optionalMemberOffset = new OptionalMemberOffset();
        optionalMemberOffset.read(reader);
        checkOptionalMemberOffset(
                optionalMemberOffset, hasOptional, OPTIONAL_FIELD_OFFSET, optionalField, field);
    }

    @Test
    public void readWithoutOptional() throws IOException, ZserioError
    {
        final boolean hasOptional = false;
        final long optionalFieldOffset = 0xABCD;
        final int field = 0x2121;
        final BitBuffer bitBuffer =
                writeOptionalMemberOffsetToBitBuffer(hasOptional, optionalFieldOffset, null, field);
        final BitStreamReader reader = new ByteArrayBitStreamReader(bitBuffer);
        final OptionalMemberOffset optionalMemberOffset = new OptionalMemberOffset();
        optionalMemberOffset.read(reader);
        checkOptionalMemberOffset(optionalMemberOffset, hasOptional, optionalFieldOffset, null, field);
    }

    @Test
    public void bitSizeOfWithOptional()
    {
        final OptionalMemberOffset optionalMemberOffset =
                new OptionalMemberOffset(true, OPTIONAL_FIELD_OFFSET, 0x1010, 0x2020);
        assertEquals(WITH_OPTIONAL_MEMBER_OFFSET_BIT_SIZE, optionalMemberOffset.bitSizeOf());
    }

    @Test
    public void bitSizeOfWithoutOptional()
    {
        final OptionalMemberOffset optionalMemberOffset = new OptionalMemberOffset(false, 0xDEAD, null, 0xBEEF);
        assertEquals(WITHOUT_OPTIONAL_MEMBER_OFFSET_BIT_SIZE, optionalMemberOffset.bitSizeOf());
    }

    @Test
    public void initializeOffsetsWithOptional()
    {
        final boolean hasOptional = true;
        final int optionalField = 0x1010;
        final int field = 0x2020;
        final OptionalMemberOffset optionalMemberOffset =
                new OptionalMemberOffset(hasOptional, WRONG_OPTIONAL_FIELD_OFFSET, optionalField, field);
        final int bitPosition = 2;
        assertEquals(WITH_OPTIONAL_MEMBER_OFFSET_BIT_SIZE, optionalMemberOffset.initializeOffsets(bitPosition));
        checkOptionalMemberOffset(
                optionalMemberOffset, hasOptional, OPTIONAL_FIELD_OFFSET, optionalField, field);
    }

    @Test
    public void initializeOffsetsWithoutOptional()
    {
        final boolean hasOptional = false;
        final long optionalFieldOffset = 0xABCD;
        final int field = 0x2020;
        final OptionalMemberOffset optionalMemberOffset =
                new OptionalMemberOffset(hasOptional, optionalFieldOffset, null, field);
        final int bitPosition = 2;
        assertEquals(WITHOUT_OPTIONAL_MEMBER_OFFSET_BIT_SIZE + bitPosition,
                optionalMemberOffset.initializeOffsets(bitPosition));
        checkOptionalMemberOffset(optionalMemberOffset, hasOptional, optionalFieldOffset, null, field);
    }

    @Test
    public void writeReadWithOptional() throws IOException, ZserioError
    {
        final boolean hasOptional = true;
        final int optionalField = 0x1A1A;
        final int field = 0xA1A1;
        final OptionalMemberOffset optionalMemberOffset =
                new OptionalMemberOffset(hasOptional, OPTIONAL_FIELD_OFFSET, optionalField, field);
        final BitBuffer bitBuffer = SerializeUtil.serialize(optionalMemberOffset);
        checkOptionalMemberOffset(
                optionalMemberOffset, hasOptional, OPTIONAL_FIELD_OFFSET, optionalField, field);
        final OptionalMemberOffset readOptionalMemberOffset =
                SerializeUtil.deserialize(OptionalMemberOffset.class, bitBuffer);
        checkOptionalMemberOffset(
                readOptionalMemberOffset, hasOptional, OPTIONAL_FIELD_OFFSET, optionalField, field);
        assertTrue(optionalMemberOffset.equals(readOptionalMemberOffset));
    }

    @Test
    public void writeWithoutOptional() throws IOException, ZserioError
    {
        final boolean hasOptional = false;
        final long optionalFieldOffset = 0xABCE;
        final int field = 0x7ACF;
        final OptionalMemberOffset optionalMemberOffset =
                new OptionalMemberOffset(hasOptional, optionalFieldOffset, null, field);
        final ByteArrayBitStreamWriter writer = new ByteArrayBitStreamWriter();
        optionalMemberOffset.write(writer);

        final BitStreamReader reader =
                new ByteArrayBitStreamReader(writer.toByteArray(), writer.getBitPosition());
        final OptionalMemberOffset readOptionalMemberOffset = new OptionalMemberOffset(reader);
        checkOptionalMemberOffset(readOptionalMemberOffset, hasOptional, optionalFieldOffset, null, field);
        assertTrue(optionalMemberOffset.equals(readOptionalMemberOffset));
    }

    private BitBuffer writeOptionalMemberOffsetToBitBuffer(
            boolean hasOptional, long optionalFieldOffset, Integer optionalField, int field) throws IOException
    {
        try (final ByteArrayBitStreamWriter writer = new ByteArrayBitStreamWriter())
        {
            writer.writeBool(hasOptional);
            writer.writeUnsignedInt(optionalFieldOffset);
            if (hasOptional)
            {
                writer.writeBits(0, 7);
                writer.writeInt(optionalField);
            }
            writer.writeInt(field);

            return new BitBuffer(writer.toByteArray(), writer.getBitPosition());
        }
    }

    private void checkOptionalMemberOffset(OptionalMemberOffset optionalMemberOffset, boolean hasOptional,
            long optionalFieldOffset, Integer optionalField, int field)
    {
        assertEquals(hasOptional, optionalMemberOffset.getHasOptional());
        assertEquals(optionalFieldOffset, optionalMemberOffset.getOptionalFieldOffset());
        if (hasOptional)
        {
            assertEquals(optionalField, optionalMemberOffset.getOptionalField());
            assertTrue(optionalMemberOffset.isOptionalFieldUsed());
        }
        else
        {
            assertFalse(optionalMemberOffset.isOptionalFieldUsed());
        }
        assertEquals(field, optionalMemberOffset.getField());
    }

    private static final int WITH_OPTIONAL_MEMBER_OFFSET_BIT_SIZE = 104;
    private static final int WITHOUT_OPTIONAL_MEMBER_OFFSET_BIT_SIZE = 65;

    private static final long OPTIONAL_FIELD_OFFSET = (long)5;
    private static final long WRONG_OPTIONAL_FIELD_OFFSET = (long)0;
}

File: test/extensions/language/offsets/python/OptionalMemberOffsetTest.py
import zserio

import Offsets


class OptionalMemberOffsetTest(Offsets.TestCase):
    def testBitSizeOfWithOptional(self):
        optionalMemberOffset = self.api.OptionalMemberOffset(True, self.OPTIONAL_FIELD_OFFSET, 0x1010, 0x2020)
        self.assertEqual(self.WITH_OPTIONAL_MEMBER_OFFSET_BIT_SIZE, optionalMemberOffset.bitsizeof())

    def testBitSizeOfWithoutOptional(self):
        optionalMemberOffset = self.api.OptionalMemberOffset(False, 0xDEAD, None, 0xBEEF)
        self.assertEqual(self.WITHOUT_OPTIONAL_MEMBER_OFFSET_BIT_SIZE, optionalMemberOffset.bitsizeof())

    def testInitializeOffsetsWithOptional(self):
        hasOptional = True
        optionalField = 0x1010
        field = 0x2020
        optionalMemberOffset = self.api.OptionalMemberOffset(
            hasOptional, self.WRONG_OPTIONAL_FIELD_OFFSET, optionalField, field
        )
        bitPosition = 2
        self.assertEqual(
            self.WITH_OPTIONAL_MEMBER_OFFSET_BIT_SIZE, optionalMemberOffset.initialize_offsets(bitPosition)
        )
        self._checkOptionalMemberOffset(
            optionalMemberOffset, hasOptional, self.OPTIONAL_FIELD_OFFSET, optionalField, field
        )

    def testInitializeOffsetsWithoutOptional(self):
        hasOptional = False
        optionalFieldOffset = 0xABCD
        field = 0x2020
        optionalMemberOffset = self.api.OptionalMemberOffset(hasOptional, optionalFieldOffset, None, field)
        bitPosition = 2
        self.assertEqual(
            self.WITHOUT_OPTIONAL_MEMBER_OFFSET_BIT_SIZE + bitPosition,
            optionalMemberOffset.initialize_offsets(bitPosition),
        )
        self._checkOptionalMemberOffset(optionalMemberOffset, hasOptional, optionalFieldOffset, None, field)

    def testReadWithOptional(self):
        hasOptional = True
        optionalField = 0x1212
        field = 0x2121
        writer = zserio.BitStreamWriter()
        OptionalMemberOffsetTest._writeOptionalMemberOffsetToStream(
            writer, hasOptional, self.OPTIONAL_FIELD_OFFSET, optionalField, field
        )
        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        optionalMemberOffset = self.api.OptionalMemberOffset.from_reader(reader)
        self._checkOptionalMemberOffset(
            optionalMemberOffset, hasOptional, self.OPTIONAL_FIELD_OFFSET, optionalField, field
        )

    def testReadWithoutOptional(self):
        hasOptional = False
        optionalFieldOffset = 0xABCD
        field = 0x2121
        writer = zserio.BitStreamWriter()
        OptionalMemberOffsetTest._writeOptionalMemberOffsetToStream(
            writer, hasOptional, optionalFieldOffset, None, field
        )
        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        optionalMemberOffset = self.api.OptionalMemberOffset.from_reader(reader)
        self._checkOptionalMemberOffset(optionalMemberOffset, hasOptional, optionalFieldOffset, None, field)

    def testWriteWithOptional(self):
        hasOptional = True
        optionalField = 0x1A1A
        field = 0xA1A1
        optionalMemberOffset = self.api.OptionalMemberOffset(
            hasOptional, self.WRONG_OPTIONAL_FIELD_OFFSET, optionalField, field
        )
        writer = zserio.BitStreamWriter()
        optionalMemberOffset.initialize_offsets(writer.bitposition)
        optionalMemberOffset.write(writer)
        self._checkOptionalMemberOffset(
            optionalMemberOffset, hasOptional, self.OPTIONAL_FIELD_OFFSET, optionalField, field
        )
        reader = zserio.BitStreamReader(writer.byte_array, writer.bitposition)
        readOptionalMemberOffset = self.api.OptionalMemberOffset.from_reader(reader)
        self._checkOptionalMemberOffset(
            readOptionalMemberOffset, hasOptional, self.OPTIONAL_FIELD_OFFSET, optionalField, field
        )
        self.assertTrue(optionalMemberOffset == readOptionalMemberOffset)

    def testWriteWithoutOptional(self):
        hasOptional = False
        optionalFieldOffset = 0xABCE
        field = 0x7ACF
        optionalMemberOffset = self.api.OptionalMemberOffset(hasOptional, optionalFieldOffset, None, field)
        bitBuffer = zserio.serialize(optionalMemberOffset)
        readOptionalMemberOffset = zserio.deserialize(self.api.OptionalMemberOffset, bitBuffer)
        self._checkOptionalMemberOffset(readOptionalMemberOffset, hasOptional, optionalFieldOffset, None, field)
        self.assertTrue(optionalMemberOffset == readOptionalMemberOffset)

    @staticmethod
    def _writeOptionalMemberOffsetToStream(writer, hasOptional, optionalFieldOffset, optionalField, field):
        writer.write_bool(hasOptional)
        writer.write_bits(optionalFieldOffset, 32)
        if hasOptional:
            writer.write_bits(0, 7)
            writer.write_signed_bits(optionalField, 32)
        writer.write_signed_bits(field, 32)

    def _checkOptionalMemberOffset(
        self, optionalMemberOffset, hasOptional, optionalFieldOffset, optionalField, field
    ):
        self.assertEqual(hasOptional, optionalMemberOffset.has_optional)
        self.assertEqual(optionalFieldOffset, optionalMemberOffset.optional_field_offset)
        if hasOptional:
            self.assertEqual(optionalField, optionalMemberOffset.optional_field)
            self.assertTrue(optionalMemberOffset.is_optional_field_used())
        else:
            self.assertFalse(optionalMemberOffset.is_optional_field_used())
        self.assertEqual(field, optionalMemberOffset.field)

    WITH_OPTIONAL_MEMBER_OFFSET_BIT_SIZE = 104
    WITHOUT_OPTIONAL_MEMBER_OFFSET_BIT_SIZE = 65

    OPTIONAL_FIELD_OFFSET = 5
    WRONG_OPTIONAL_FIELD_OFFSET = 0
